pkg/k8s/watchers/endpoint.go

```diff
@@ -85,10 +88,90 @@ func (k *K8sWatcher) updateK8sEndpointV1(oldEP, newEP *slim_corev1.Endpoints, sw
 	if option.Config.BGPAnnounceLBIP {
 		k.bgpSpeakerManager.OnUpdateEndpoints(newEP)
 	}
+	k.addKubeAPIServerServiceEPs(newEP)
 	return nil
 }
 
 func (k *K8sWatcher) deleteK8sEndpointV1(ep *slim_corev1.Endpoints, swg *lock.StoppableWaitGroup) error {
 	k.K8sSvcCache.DeleteEndpoints(ep, swg)
 	return nil
 }
+
+// handleKubeAPIServerServiceEPChanges associates the set of 'desiredIPs' with
+// the 'reserved:kube-apiserver' label in the IPCache. This allows policy
+// selectors for the kube-apiserver entity to match these peers.
+//
+// Any IPs currently associated with the apiserver that are not part of
+// 'desiredIPs' will be disassociated from the apiserver following a call to
+// this function.
+//
+// The actual implementation of this logic down to the datapath is handled
+// asynchronously.
+func (k *K8sWatcher) handleKubeAPIServerServiceEPChanges(desiredIPs map[string]struct{}) {
+	// We must perform a diff on the ipcache.IdentityMetadata map in order to
+	// figure out which IPs are stale and should be removed, before we inject
+	// new IPs into the ipcache. The reason is because kube-apiserver will
+	// constantly reconcile this specific object, even when it's been deleted;
+	// effectively, this means we can avoid listening for the delete event.
+	// Therefore, any changes to this specific object can be handled in a
+	// "flattened" manner, since the most up-to-date form of it will be an add
+	// or update event. The former is sent when Cilium is syncing with K8s and
+	// the latter is sent anytime after.
+	//
+	// For example:
+	//   * if a backend is removed or updated, then this will be in the form of
+	//     an update event.
+	//   * if the entire object is deleted, then it will quickly be recreated
+	//     and this will be in the form of an add event.
+	ips := ipcache.FilterMetadataByLabels(labels.LabelKubeAPIServer)
+	oldAPIServerIPs := make(map[string]struct{}, len(ips))
+	for _, v := range ips {
+		oldAPIServerIPs[v] = struct{}{}
+	}
+
+	toRemove := make(map[string]labels.Labels)
+	for ip := range oldAPIServerIPs {
+		if _, ok := desiredIPs[ip]; !ok {
+			toRemove[ip] = labels.LabelKubeAPIServer
+		}
+	}
```

Comment:
I think the first map can be avoided and the code simplified. I'm guessing the goal was to avoid duplicated IPs via the `oldAPIServerIPs` set, but since we're adding into the `toRemove` map afterward, is that really a concern?
```suggestion
	oldAPIServerIPs := ipcache.FilterMetadataByLabels(labels.LabelKubeAPIServer)
	toRemove := make(map[string]labels.Labels)
	for _, ip := range oldAPIServerIPs {
		if _, ok := desiredIPs[ip]; !ok {
			toRemove[ip] = labels.LabelKubeAPIServer
		}
	}
```

Reply:
When I first reviewed this, I assumed it was because `FilterMetadataByLabels` returns a slice and we needed a map to do the comparison correctly. Also previously the `desiredIPs` map was populated in between these two calls. But looking back over I think you're completely right, we only iterate, we don't need indexing. This will be a nice tidyup. :+1: